Support for Arnold Operators on Softimage RenderPasses

caron opened the pull request:
This implements support for Arnold Operators on RenderPasses. It uses the OutputShaderStack and a custom dummy node named 'operator'. This node has no C++ shader code it is implemented only in jscript. This node is skipped during exporting of the operator and since AOV Shaders use the OutputShaderStack too, we skip the operator there also.

caron committed:
operator nodes are working with shader auto-generation now

caron committed:
fixes from code review

caron committed:
move operator metadata into its own file

JenusL committed:
move pretty label to own function in tools

JenusL committed:
change operators metadata

JenusL committed:
support shader parameter arrays

JenusL committed:
change operator to ref type

JenusL committed:
small cleanup

JenusL committed:
remove more uses of operator type

JenusL committed:
add warning if unknown referencefiltertype

caron committed:
Merge pull request #1 from JenusL/feat/shader_array_defs

Feat/shader array defs

caron committed:
remove custom port type registration

caron committed:
Adding operator related switches to render options

caron committed:
Support for outputting operators to ass files.

caron committed:
adjust include order to match other files

caron committed:
adjust function signature to include frame

caron committed:
need inputs param to be inspectable

trying to reuse existing shader parsing code which would skip params which are not inspectable

caron committed:
trying to use 'LoadShader' function to parse operator graph

for siShaderArrayParameterID param type added support for arnold arrays of node type

caron committed:
working on IPR updates for operators

caron committed:
refined LoadShaderParameter function with better support for node arrays

since we are reusing shader parsing functions the operator's 'inputs' parameter was using the wrong method for node arrays, the existing code would call AiNodeLink on each array item. this was causing arnold to spit back warnings about not being able to link the nodes.
note about this implementation... the array is allocated and set prior to parsing the array values, this means we need to retrieve the array at a later time once the value of the array item is known. then we can set that value in the array and set it on the node again.

caron committed:
implement operators connected to the output shader stack

this is an alternative method for implementing operators which will use a dummy shader connected to the output shader stack. this works better with IPR

caron committed:
behavior of GetConnectedShader wasn't working well for this case

GetConnectedShader function wants to return the node that owns the parameter if nothing is connected. in this case we never want to export the dummy operator shader node.

caron committed:
removed custom port from the pass

so bummed this didn't work!

caron committed:
new operator node added in arnold 5.3, 'include_graph'

caron commented:
odd to see all the history of comments on my commits out of order.

anyways this work addresses Autodesk/sitoa#19

JenusL commented on plugins/sitoa/renderer/IprShader.cpp:
```diff
@@ -309,6 +309,9 @@ void UpdatePassShaderStack(const Pass &in_pass, double in_frame)
          if (outputShader.IsValid())
          {
+            // skip 'operator' dummy shader
+            if (outputShader.GetName() == L"operator")
```
Are you sure this isn't just getting the name of that shader node instance. That is if you copy it, it will be named "operator1"?
Could you try that just to make sure it works?
If your code dosn't work I recommend you use the code that is used everywhere else in SItoA:
```suggestion
            if (GetShaderNameFromProgId(outputShader.GetProgID()) == L"operator")
```

caron replied:
Thanks for catching that, I will switch to the GetShaderNameFromProgId.

caron committed:
use the proper method for checking shader type

was stoopidly just checking the name and it failed as soon as you renamed the node.

JenusL commented:
So I'm getting ready to merge this. Will go over the code one more time first.
@caron Did you see my question about replacing closure node with a new one that has both a closure and operator input. That way they could be added to Procedurals as well.
Another alternative would be to add a material node Picker on the Arnold Procedural PPG to pick a material that has the dummy operator node you created.
That might actually be better and less confusing.

JenusL commented on plugins/sitoa/common/Tools.h:
```diff
@@ -89,16 +89,17 @@ namespace ATSTRING
-   //
+   // sitoa custom ports/types
    const AtString closure("closure");
+   const AtString operator_node("operator");
```
Is this used now or could it be removed?
```suggestion
```

caron replied:
Ya, this should be removed, I was having trouble getting a scripted shader plugin working and thought it needed to have a c++ counter part. This is left over from that experiment...

caron replied:
Actually, this is a remnant from when we started implementing shader def support for operators. Still, I did a quick find in files for the entire code base and didn't see anywhere it was used.

caron committed:
unused string constant

caron commented:
@JenusL , so following up about deprecating the 'closure' node... I think it is up to you and the community. I don't use SItoA in production anymore so I am not so worried about old scenes and old setups. But I could imagine the change in behavior could cause some headaches.

The material node picker option sounds nice but my first thought would be if someone wanted to mix the material override and the operator on a procedural. We would still need a node which had both `closure` and `operator` ports, right?

JenusL commented:
No we wont need a node with both ports if one can be assigned on the procedural object (standard material behavior) and then another node picked in the ppg (operator node).
So I will leave everything as it is now and maybe later att a Operatro Material Picker in Procedural PPG. But that's for later if ever :)

caron commented:
So just a string parameter with a path to a material that has the `operator` on it?

If that is the case then my concern is having materials floating around which aren't connected. They end up getting cleaned up easily by built-in tools (remove unused materials) and custom pipeline tools.

JenusL commented:
That might be true, but I would probably add a naming prefix to my floating "materials" to indicate that it's an operator. For example, an operator made to override all materials in the scene I would maybe call `op_OverrideMaterial`.
Then that custom tool just have to ignore materials that start with `op_`.
Ugly, I know,
The only other way I see is some kind of joint closure/operator dummy node that I mentioned earlier.

caron commented:
That doesn't stop people from running 'Remove Unused Materials' and blowing away operators that are supposed to be connected to Procedurals.

I am leaning towards a new dummy node. no reason to remove support for the existing `closure` and `operator` dummy nodes.

JenusL commented:
Ok. And I totally forgot about IPR as well. That wouldn't work with my PPG idea and floating materials.
Since we haven't released any version of SItoA with the `operator` node, I might remove it if I find it to be redundant. Is that OK?
What would be a good name for this node? Simply `arnold` or maybe `arnold_glue`?

caron commented:
Before I say 'yes'... we should talk through how the new node would work on the pass operator. Since we had to put it in the Output Shader Stack it was easy to skip by asking for the connected node progid. If the `closure` and `operator` ports are merged then we need to switch to checking the ports themselves.

new node name? ugh... I don't know :(

`arnold_ports`
`arnold_root`
`arnold_terminal`

JenusL commented:
Yeah I already thought about that we need to check the ports instead of the shader.
Naming stuff is the hardest 😂
But I kinda like `arnold_ports`

caron commented:
Again, I don't use SItoA in production anymore so I will leave the decision up to you.
